students/erica_winberry/session04/mailroom2.py

```
@@ -0,0 +1,188 @@
+and add some tests
+and list (and dict, and set) comprehensions...
+"""
+
+donors = [
+    {"name": "Carol Danvers", "donations": [25.00, 100.00]},
+    {"name": "Kumala Khan", "donations": [15.00, 15.00, 25.00]},
+    {"name": "Jennifer Walters", "donations": [50.00, 100.00, 65.00]},
+    {"name": "Monica Rambeau", "donations": [200.00]},
+    {"name": "Jessica Drew", "donations": [45.00, 30.00, 70.00]}
+    ]
+
+
+def safe_input(message):
+    # Prevents error when ctrl-c or ctrl-d are used to stop the script.
+    try:
+        user_input = input(message)
+        return user_input
+    except (KeyboardInterrupt, EOFError):
+        return None
+
+
+def intro():
+    print("\nMAILROOM OPTIONS")
+    print("1 - Send a Thank You")
+    print("2 - Create a Report")
+    print("3 - Write Thank You letters to all donors.")
+    print("\n(Type 'exit' at any time.)")
+
+
+def thanks():
+    # Collects information from the user when sending a thank you letter.
+    donor_name = get_name(donors)
+    if donor_name == "exit":
+        return "Returning to main menu.\n"
+    donation = get_amount(donor_name, donors)
+    if donation == "exit":
+        return "Returning to main menu.\n"
+    elif donation != 0:
+        write_letter(donor_name, donation)
+    else:
+        print("Returning to main menu.\n")
+
+
+def report(report_source):
+    # Print a list of your donor, sorted by total historical donation amount.
+    print("{:^80}".format("DONOR REPORT\n"))
+    print("{:>20}{:>20}{:>20}{:>20}".format("Donor Name", "Total Amt. \
+Donated", "Times Donated", "Avg. Donation"))
+    print("{:>20}".format("------------------") * 4)
+    for donor_dictionary in report_source:
+        donor = donor_dictionary["name"]
+        total_amount = sum(donor_dictionary["donations"])
+        total_donations = len(donor_dictionary["donations"])
+        try:
+            avg_donation = total_amount / total_donations
+        except ZeroDivisionError as e:
+            return e
```

you generally do'nt want to return the Exception object -- you want to handle the exception -- in this case, I'd say set the average_donation to 0.0 -- then that's it.

If you do something, but still need the error propogate, then you re-raise it:

raise e